storage/src/main/java/org/apache/kafka/storage/internals/log/UnifiedLog.java

```diff
@@ -55,6 +113,2302 @@ public class UnifiedLog {
+                                deleteProducerSnapshots(deletedSegments, true);
+                                leaderEpochCache.truncateFromEndAsyncFlush(targetOffset);
+                                logStartOffset = Math.min(targetOffset, logStartOffset);
+                                rebuildProducerState(targetOffset, producerStateManager);
+                                if (highWatermark() >= localLog.logEndOffset())
+                                    updateHighWatermark(localLog.logEndOffsetMetadata());
+                            }
+                            return true;
+                        }
+                    }
+                });
+    }
+
+    /**
+     *  Delete all data in the log and start at the new offset
+     *
+     *  @param newOffset The new offset to start the log with
+     *  @param logStartOffsetOpt The log start offset to set for the log. If None, the new offset will be used.
+     */
+    public void truncateFullyAndStartAt(long newOffset, Optional<Long> logStartOffsetOpt) {
+        maybeHandleIOException(
+                "Error while truncating the entire log for " + topicPartition() + " in dir " + dir().getParent(),
+                () -> {
+                    logger.debug("Truncate and start at offset {}, logStartOffset: {}", newOffset, logStartOffsetOpt.orElse(newOffset));
+                    synchronized (lock)  {
+                        localLog.truncateFullyAndStartAt(newOffset);
+                        leaderEpochCache.clearAndFlush();
+                        producerStateManager.truncateFullyAndStartAt(newOffset);
+                        logStartOffset = logStartOffsetOpt.orElse(newOffset);
+                        if (remoteLogEnabled()) localLogStartOffset = newOffset;
+                        rebuildProducerState(newOffset, producerStateManager);
+                        return updateHighWatermark(localLog.logEndOffsetMetadata());
+                    }
+                });
+    }
+
+    /**
+     * The time this log is last known to have been fully flushed to disk
+     */
+    public long lastFlushTime() {
+        return localLog.lastFlushTime();
+    }
+
+    /**
+     * The active segment that is currently taking appends
+     */
+    public LogSegment activeSegment() {
+        return localLog.segments().activeSegment();
+    }
+
+    /**
+     * All the log segments in this log ordered from oldest to newest
+     */
+    public Collection<LogSegment> logSegments() {
+        return localLog.segments().values();
+    }
+
+    /**
+     * Get all segments beginning with the segment that includes "from" and ending with the segment
+     * that includes up to "to-1" or the end of the log (if to > logEndOffset).
+     */
+    public Collection<LogSegment> logSegments(long from, long to) {
+        synchronized (lock) {
+            return localLog.segments().values(from, to);
```

Comment:
Could we create a copy while holding the lock and return copied value? Ditto for `logSegments()` and `nonActiveLogSegmentsFrom()`.

Reply:
I've updated these 3 methods to return a copy of the collection.